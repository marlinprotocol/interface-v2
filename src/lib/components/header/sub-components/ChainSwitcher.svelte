<script lang="ts">
	import { buttonClasses } from '$lib/atoms/componentClasses';
	import { switchChain } from '$lib/utils/helpers/networkHelper';
	import { environment } from '$lib/data-stores/environment';
	import { staticImages } from '$lib/components/images/staticImages';
	import { connected, web3WalletStore } from '$lib/data-stores/walletProviderStore';
	import {
		allowedChainsStore,
		chainStore,
		updateChainStore
	} from '$lib/data-stores/chainProviderStore';
	import Icon from '$lib/atoms/icons/Icon.svelte';
	import chevronDown from 'svelte-awesome/icons/chevronDown';

	export let id = 'chain-dropdown';

	let activeChainImage = staticImages.chainLogo;
	$: dropdownOpen = false;

	function handleChainSwitch(chainId: number) {
		if ($connected && chainId !== $chainStore.chainId) {
			switchChain(chainId, $web3WalletStore[0].provider);
		} else {
			updateChainStore(chainId);
		}
		dropdownOpen = false; // Close dropdown after switching
	}

	function toggleDropdown() {
		dropdownOpen = !dropdownOpen;
	}

	function closeDropdown() {
		dropdownOpen = false;
	}

	function closeSwitcherWhenClickedOutside(event: MouseEvent) {
		const dropdown = document.getElementById(id);
		if (!dropdown?.contains(event.target as Node)) {
			closeDropdown();
		}
	}

	$: {
		const activeChainId = $chainStore.chainId;
		if (activeChainId !== null) {
			activeChainImage = environment.valid_chains[activeChainId]?.chain_image;
		}
	}
</script>

<svelte:window on:click={(e) => closeSwitcherWhenClickedOutside(e)} />

<div class="dropdown dropdown-end" {id}>
	<!-- svelte-ignore a11y-click-events-have-key-events -->
	<div
		tabindex={$allowedChainsStore.length === 0 ? -1 : 0}
		role="button"
		class="{buttonClasses.whiteFilled} chain-btn m-1 flex h-12"
		on:click={toggleDropdown}
	>
		<div class="flex items-center justify-center">
			<div class="flex h-8 w-8 items-center justify-center">
				{#if activeChainImage}
					<img src={activeChainImage} alt="current chain" />
				{:else}
					<span class="flex h-8 w-8 items-center justify-center rounded-full bg-primary text-white">
						{$chainStore.chainDisplayName.charAt(0).toLocaleUpperCase()}
					</span>
				{/if}
			</div>
			<div class="ml-2">
				<Icon data={chevronDown} size={10} iconColorClass="icon-primary" />
			</div>
		</div>
	</div>
	{#if dropdownOpen}
		<!-- svelte-ignore a11y-no-noninteractive-tabindex -->
		<ul
			tabindex={$allowedChainsStore.length === 0 ? -1 : 0}
			class="menu dropdown-content rounded-box z-[10] mt-4 w-52 bg-base-100 p-2 shadow"
		>
			{#each $allowedChainsStore as chain (chain)}
				<li class="flex {$chainStore.chainId === chain ? 'rounded-lg bg-primary text-white' : ''}">
					<button on:click={() => handleChainSwitch(chain)}>
						<div class="h-6 w-6 rounded-full ring-1 ring-white">
							<img src={environment.valid_chains[chain].chain_image} alt="Chain Logo" />
						</div>
						{environment.valid_chains[chain].chain_name}
					</button>
				</li>
			{/each}
		</ul>
	{/if}
</div>
