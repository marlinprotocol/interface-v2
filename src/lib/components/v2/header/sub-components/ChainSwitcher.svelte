<script lang="ts">
	import { buttonClasses } from '$lib/atoms/componentClasses';
	import { switchChain } from '$lib/utils/helpers/networkHelper';
	import { environment } from '$lib/data-stores/environment';
	import { staticImages } from '$lib/components/images/staticImages';
	import { connected, web3WalletStore } from '$lib/data-stores/walletProviderStore';
	import {
		allowedChainsStore,
		chainStore,
		updateChainStore
	} from '$lib/data-stores/chainProviderStore';
	import Icon from '$lib/atoms/icons/Icon.svelte';
	import chevronDown from 'svelte-awesome/icons/chevronDown';

	let activeChainImage = staticImages.chainLogo;

	function handleChainSwitch(chainId: number) {
		if ($connected && chainId !== $chainStore.chainId) {
			switchChain(chainId, $web3WalletStore[0].provider);
		} else {
			updateChainStore(chainId);
		}
	}

	function closeSwitcherWhenClickedOutside(event: MouseEvent) {
		const dropdown = document.getElementById('chain-dropdown');
		if (!dropdown?.contains(event.target as Node)) {
			dropdown?.removeAttribute('open');
		}
	}

	$: {
		const activeChainId = $chainStore.chainId;
		if (activeChainId !== null) {
			activeChainImage = environment.valid_chains[activeChainId]?.chain_image;
		}
	}
</script>

<svelte:window on:click={(e) => closeSwitcherWhenClickedOutside(e)} />

<div class="dropdown dropdown-end" id="chain-dropdown">
	<div
		tabindex={$allowedChainsStore.length === 0 ? -1 : 0}
		role="button"
		class="{buttonClasses.whiteFilled} chain-btn m-1 flex h-12"
	>
		<div class="flex items-center justify-center">
			<div class="h-8 w-8">
				<img src={activeChainImage} alt="current chain" />
			</div>
			<div class="ml-2">
				<Icon data={chevronDown} size={10} iconColorClass="icon-primary" />
			</div>
		</div>
	</div>
	<!-- svelte-ignore a11y-no-noninteractive-tabindex -->
	<ul
		tabindex={$allowedChainsStore.length === 0 ? -1 : 0}
		class="menu dropdown-content rounded-box z-[1] mt-4 w-52 bg-base-100 p-2 shadow"
	>
		{#each $allowedChainsStore as chain (chain)}
			<li class="flex {$chainStore.chainId === chain ? 'rounded-lg bg-primary text-white' : ''}">
				<button on:click={() => handleChainSwitch(chain)}>
					<div class="h-6 w-6 rounded-full ring-1 ring-white">
						<img src={environment.valid_chains[chain].chain_image} alt="Chain Logo" />
					</div>
					{environment.valid_chains[chain].chain_name}
				</button>
			</li>
		{/each}
	</ul>
</div>
