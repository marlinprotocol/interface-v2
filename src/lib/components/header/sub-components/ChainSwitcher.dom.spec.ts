import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/svelte';
import ChainSwitcher from './ChainSwitcher.svelte';

const {
	web3WalletStore,
	walletStore,
	chainStore,
	chainConfigStore,
	connected,
	allowedChainsStore
} = await vi.hoisted(() => import('$lib/data-stores/mock-data-stores/mockStores'));

beforeAll(() => {
	vi.mock('$lib/data-stores/walletProviderStore', async () => {
		return {
			web3WalletStore,
			walletStore,
			connected
		};
	});
	vi.mock('$lib/data-stores/chainProviderStore', async () => {
		return {
			chainStore,
			chainConfigStore,
			allowedChainsStore
		};
	});
});

describe('ChainSwitcher', () => {
	it("render's the component properly", async () => {
		const { container } = render(ChainSwitcher);
		expect(container).toMatchSnapshot();
	});

	it('displays the correct chain image', async () => {
		chainStore.mockSetSubscribeValue({
			chainDisplayName: '',
			chainId: 43,
			chainImage: '/images/arbitrum.webp',
			chainName: '',
			isValidChain: true
		});

		const { getByAltText, findByRole } = render(ChainSwitcher);
		const button = await findByRole('button');
		await button.click(); // Wait for the button and then click it

		const chainLogo = getByAltText('Chain Logo');
		expect(chainLogo.getAttribute('src')).toBe('/images/arbitrum.webp');
	});
});
