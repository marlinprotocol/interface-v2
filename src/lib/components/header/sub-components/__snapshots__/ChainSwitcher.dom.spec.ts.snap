// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChainSwitcher > render's the component properly 1`] = `
<body>
  <onboard-v2
    style="all: initial;"
  />
  <style>
    
    
@font-face {
  font-family: 'Inter';
  font-style:  normal;
  font-weight: 300 600;
  font-display: swap;
  src: url("https://rsms.me/inter/font-files/Inter-roman.var.woff2?v=3.19") format("woff2");
}

  
  </style>
  <div>
    <div
      class="dropdown dropdown-end"
      id="chain-dropdown"
    >
      <div
        class="btn btn-theme btn-sm items-center flex justify-center disabled:cursor-not-allowed !rounded-[100px] bg-white text-black border border-[#D9DADE] hover:bg-base-200 chain-btn m-1 flex h-12"
        role="button"
        tabindex="0"
      >
        <div
          class="flex items-center justify-center"
        >
          <div
            class="flex h-8 w-8 items-center justify-center"
          >
            <img
              alt="current chain"
              src="/images/arbitrum.webp"
            />
          </div>
           
          <div
            class="ml-2"
          >
            <svg
              aria-label=""
              class="fa-icon icon-primary s-T5YyQewRBd7n"
              data-testid="icon"
              height="16"
              role="presentation"
              style="width:10px; height:10px; cursor:pointer;  "
              version="1.1"
              viewBox="0 0 1792 1792"
              width="16"
            >
              <path
                d="M1683 808l-742 741q-19 19-45 19t-45-19l-742-741q-19-19-19-45.5t19-45.5l166-165q19-19 45-19t45 19l531 531 531-531q19-19 45-19t45 19l166 165q19 19 19 45.5t-19 45.5z"
              />
              
               
               
              
              
            </svg>
            <!--&lt;Svg&gt;-->
            <!--&lt;Icon&gt;-->
            <!--&lt;Icon&gt;-->
          </div>
        </div>
      </div>
       
      <ul
        class="menu dropdown-content rounded-box z-[1] mt-4 w-52 bg-base-100 p-2 shadow"
        tabindex="0"
      >
        <li
          class="flex rounded-lg bg-primary text-white"
        >
          <button>
            <div
              class="h-6 w-6 rounded-full ring-1 ring-white"
            >
              <img
                alt="Chain Logo"
                src="/images/arbitrum.webp"
              />
            </div>
             
            Arbitrum Sepolia
          </button>
           
        </li>
      </ul>
    </div>
    <!--&lt;ChainSwitcher&gt;-->
  </div>
</body>
`;
